src/item.cpp

```diff
@@ -3949,21 +4015,36 @@ bool item::ammo_consume( int qty ) {
         return false;
     }
 
+    item *mag = magazine_current();
+    if( mag ) {
+        auto res = mag->ammo_consume( qty );
+        if( res && ammo_remaining() == 0 && mag->has_flag( "NO_RELOAD" ) ) {
+            contents.erase( std::remove_if( contents.begin(), contents.end(), [&mag]( const item& e ) {
+                return mag == &e;
+            } ) );
+        }
+        return res;
+    }
+
     if( qty > ammo_remaining() ) {
         return false;
     }
 
-    if( is_tool() ) {
-        charges -= qty;
-        if( charges == 0 ) {
-            unset_curammo();
+    if( is_magazine() ) {
+        while( contents.size() ) {
+            auto& e = *contents.rbegin();
```

Comment:
Btw. to get a reference to the last item in a vector, you can use `std::vector::back()` (also works for lists).